### ISSUE TITLE
:four_leaf_clover: `Proposal`: Improving The Robustness of Config File

### ISSUE THREAD

**Issue by daniel-hutao** [devstream-io/devstream#659]
## What Would You Like to Add? Why Is This Needed?

After #655 merged, se now support single config file like below:

https://github.com/devstream-io/devstream/blob/221298b5d5dde98a67ebadb99197fcb89218f7b0/examples/gitops.yaml

As you can see, there is an order requirement here, e.g.
1. core config
2. var config
3. tool config

What if the user does not write in this order? I want dtm to still work properly.

## Design

All three types of configurations in any order can be successfully identified. Remember to test fully to ensure that all these scenarios work properly:

1. core config only.
2. core config + tool config / tool config + core config
3. core config + tool config + var config / ... (any order)

## Anything else

Take the first step, go for it! Feel free to contact me if you have any difficulties.

**Comment by byene0923** [comment]
sounds interesting, i want to have a try

**Comment by aFlyBird0** [comment]
> sounds interesting, i want to have a try

Welcome to DevStream! Please enjoy writing code!

**Comment by daniel-hutao** [comment]
@byene0923 happy coding！

**Comment by byene0923** [comment]
Hi, I think the core problem is to distinguish core config, var config and tool config. so we can add some limit to the config
- core config must exist and be unique, the key `varFile` and `toolFile ` must exist and only in core config. we use one of them to identify the core config
- tool config may be nil or only have one(multiple?). the key `tools` must exist and only in tool config, we use it identify the tool config
- var config is `map[string]interface{}` and can be multiple or nil.  the key must not be `varFile`, `toolFile`, `tools` and i think the key is unique

a valid yaml may be as
```yaml
---
# plugins config
...
---
# variables config
...
---
# core config
....
---
# variables config
...
```

**Comment by aFlyBird0** [comment]
> Hi, I think the core problem is to distinguish core config, var config and tool config. so we can add some limit to the config
> 
> * core config must exist and be unique, the key `state` must exist and only in core config. we use it to identify the core config
> * tool config may be nil or only have one(multiple?). the key `tools` must exist and only in tool config, we use it identify the tool config
> * var config is `map[string]interface{}` and can be multiple or nil.  the key must not be `state`, `tools` and i think the key is unique
> 
> a valid yaml may be as
> 
> ```yaml
> ---
> # plugins config
> ...
> ---
> # variables config
> ...
> ---
> # core config
> ....
> ---
> # variables config
> ...
> ```
> 
> or is there any way to distinguish them? may use validaition?

Good job! Your solution is very good and clearly expressed.

1. I agree with you totally for the first idea. I think it is the best way.
2. For the second idea/question about tool config: I think there must be only one tool config, not nil and not multiple. But it can be defined in main config file by the key `tools` or in `toolFile`. Yep, we can regard the config block surrounded by `---` as `tool config` if it has the key `tools`.
3. For the third idea about var config: Sorry, I can't understand all of them. Do you mean if a config block is neither core config nor tool config, it should be var config? And you think var config should be the type of `map[string]interface{}`? 

Besides, my previous idea and solution is almost the same as yours.

Here's a minor detail, in your sample config file, the var file can be split into two sections. This is very robust, but I don't know if devstream plans to make it that way, maybe it agrees that the var config must be in the same block.

What's your opinion? @daniel-hutao @IronCore864

**Comment by byene0923** [comment]
> For the third idea about var config: Sorry, I can't understand all of them. Do you mean if a config block is neither core config nor tool config, it should be var config? And you think var config should be the type of map[string]interface{}?

Thanks for your resp,  I see varConfig is[ Unmarshaled into map](https://github.com/devstream-io/devstream/blob/main/internal/pkg/configloader/toolconfig.go#L113)

> Here's a minor detail, in your sample config file, the var file can be split into two sections. This is very robust, but I don't know if devstream plans to make it that way, maybe it agrees that the var config must be in the same block.


yeh  i think varconfig is flexible， may not be in one block。 and more generic, may there are different config types in the future ? we may think not limited by this three types

**Comment by aFlyBird0** [comment]
I agree with you. You did a perfect job! Let's wait for PMC's opinion of dtm. @byene0923

**Comment by aFlyBird0** [comment]
@byene0923 Hello!

After discussing with Daniel Hu, PMC of devstream here are our opinions.

Your solution is robust but we can do it more simply in this issue. The goal of `good first issue` is to help new contributors finish their first contribution fluently. 

That is to say, you need not support multiple sections of var config. It is complex and may cause a mass. The core logic of var config is: Reading var file if `varFIle` exists, or reading from main config yaml. We just support a single section of var config and keep it simple. Covenant over configuration.

The goal of this issue is just to support any order of different configs. It's enough. If you want to do more, you can submit a new issue to discuss.

As for your first idea and second idea, I totally support you! Happy coding!

**Comment by byene0923** [comment]
sure，I'm a little busy in work these days
